<a href="https://colab.research.google.com/github/JapiKredi/Pinnacle_AI_program_AnalyticsVidyha/blob/main/Pandas_SqllikejoinsinPython.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---
---

<center><h1> 📍 📍 Working with Multiple DataFrames 📍 📍 </h1></center>


---

- In most of the real life projects you will not get data from a single resource. You might need to combine data that you gather from multiple sources. In this notebook, we will see how to work with multiple data files?


***We have already worked with big mart sales data. Here, we have divided the data based on different outlet size. Let's see how to work with the multiple files***


---

#### `READ THE DATA`

---

In [1]:
# importing the pandas library
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


---

***Read 3 different files `outlet_size_small.csv`, `outlet_size_medium.csv` and `outlet_size_high.csv` stored in the folder datasets.***

---

In [4]:
# read the datasets
outlet_size_small = pd.read_csv('/content/drive/My Drive/Pandas_Concatenation/Concatenation data/outlet_size_small.csv')
outlet_size_medium = pd.read_csv('/content/drive/My Drive/Pandas_Concatenation/Concatenation data/outlet_size_medium.csv')
outlet_size_large = pd.read_csv('/content/drive/My Drive/Pandas_Concatenation/Concatenation data/outlet_size_high.csv')

***Let's check the shape of the data.***

---

In [5]:
outlet_size_small.shape, outlet_size_medium.shape, outlet_size_large.shape

((2388, 9), (2793, 9), (932, 9))

***So, there are 2388 small, 2793 medium and 932 large size outlets are there.***


---

***Let's have a look at the data***

#### `OUTLET SIZE SMALL`
---

In [6]:
outlet_size_small.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  
0  144.1102            OUT046  Supermarket Type1          2187.1530  
1  119.6782            OUT046  Supermarket Type1          2145.2076  
2   56.3614            OUT046  Supermarket Type1          1547.3192  
3  230.5352            OUT035  Supermarket Type1          2748.4224  
4  196.5084            OUT035  Supermarket Type1          1587.2672

#### `OUTLET SIZE MEDIUM`

---

In [7]:
outlet_size_medium.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15        9.300          Low Fat         0.016047   
1           DRC01        5.920          Regular         0.019278   
2           FDN15       17.500          Low Fat         0.016760   
3           FDP36       10.395          Regular         0.000000   
4           FDP10          NaN          Low Fat         0.127470   

      Item_Type  Item_MRP Outlet_Identifier        Outlet_Type  \
0         Dairy  249.8092            OUT049  Supermarket Type1   
1   Soft Drinks   48.2692            OUT018  Supermarket Type2   
2          Meat  141.6180            OUT049  Supermarket Type1   
3  Baking Goods   51.4008            OUT018  Supermarket Type2   
4   Snack Foods  107.7622            OUT027  Supermarket Type3   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           556.6088  
4          4022.7636

#### `OUTLET SIZE LARGE`

---

In [8]:
outlet_size_large.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           NCD19         8.93          Low Fat         0.000000   
1           FDO10        13.65          Regular         0.012741   
2           FDF32        16.35          Low Fat         0.068024   
3           FDN22        18.85          Regular         0.138190   
4           DRJ59        11.65          low fat         0.019356   

               Item_Type  Item_MRP Outlet_Identifier        Outlet_Type  \
0              Household   53.8614            OUT013  Supermarket Type1   
1            Snack Foods   57.6588            OUT013  Supermarket Type1   
2  Fruits and Vegetables  196.4426            OUT013  Supermarket Type1   
3            Snack Foods  250.8724            OUT013  Supermarket Type1   
4            Hard Drinks   39.1164            OUT013  Supermarket Type1   

   Item_Outlet_Sales  
0           994.7052  
1           343.5528  
2          1977.4260  
3          3775.0860  
4           308.9312

---

#### `CONCATENATE ALL THE DATAFRAMES`


We will use the concat function to concatenate all the dataframes. You just need to pass the list of dataframes to concatenate.

---

#### `FOR ROW-WISE CONCATENATION USE AXIS=0`


---

In [9]:
# dataframes list
all_dataframes = [outlet_size_small, outlet_size_medium, outlet_size_large]

In [10]:
# concatenate all the dataframes
data = pd.concat(all_dataframes, axis=0)

In [11]:
# shape of the data
data.shape

(6113, 9)

#### `FOR COLUMN-WISE CONCATENATION USE AXIS=1`


It is not advised to concatenate dataframes column wise. If you want to then you need to take care of some checks like the number of rows must be same in both the dataframes. Indexes are sorted of both the dataframes. If you are done with all the checks then you can simply use `axis=1` to do the job.



Let's see with the help of an example.


---

In [12]:
sample_dataframe = pd.DataFrame({
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})
sample_dataframe

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

***Let's create a dataframe which contains column name `phone_no`. Here, we are assuming that order of phone numbers are correct as the order of names in the sample_dataframe.***

---

In [13]:
# another sample dataframe
phone_no = pd.DataFrame({ 'phone_no' : [212202, 202021, 212334, 213431, 211721]})
phone_no

phone_no
0    212202
1    202021
2    212334
3    213431
4    211721

In [14]:
combined = pd.concat([sample_dataframe,phone_no], axis=1)
combined

roll_no     name grade  marks       city  phone_no
0      102  Aravind     B     15   Gurugram    212202
1      101    Rahul     B     15      Delhi    202021
2      104  Prateek     A     20      Delhi    212334
3      103  Piyuesh     C      4   Gurugram    213431
4      105   Kartik     A     22  Hyderabad    211721

---

#### `Performing SQL-Like Joins in Pandas`


#### `LET'S LEARN ABOUT THE JOINS IN PANDAS WITH SOME EXAMPLES`

- We will create a sample dataframe of students data that will contain `roll_no`, `name`, `grade`, `marks` and `city`.


---

In [15]:
student_df = pd.DataFrame({
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})
student_df

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

---

***Now, we have a mapping of city and state. We want to add another feature to our dataframe state using this mapping.***


Let's create the city_state_mapping

---

In [16]:
city_state_mapping = pd.DataFrame({
    'city' :  ['Gurugram', 'Delhi', 'Hyderabad', 'Faridabad'],
    'state' : ['Haryana',  'Delhi', 'Telangana', 'Haryana']
})
city_state_mapping

city      state
0   Gurugram    Haryana
1      Delhi      Delhi
2  Hyderabad  Telangana
3  Faridabad    Haryana

***Now, we want to add another column state to the `student_df` using the `city_state_mapping`. We can do this by doing a left join. We need to use the merge function and set the parameters `how='left'` and `on='city`.***

![](left.png)
---

In [17]:
city_state_mapping

city      state
0   Gurugram    Haryana
1      Delhi      Delhi
2  Hyderabad  Telangana
3  Faridabad    Haryana

In [20]:
student_df.merge(city_state_mapping, how='left', on='city')

roll_no     name grade  marks       city      state
0      102  Aravind     B     15   Gurugram    Haryana
1      101    Rahul     B     15      Delhi      Delhi
2      104  Prateek     A     20      Delhi      Delhi
3      103  Piyuesh     C      4   Gurugram    Haryana
4      105   Kartik     A     22  Hyderabad  Telangana

***Now, we have another dataframe that contains roll_no of some students. We need to find out the other details of the students. We can do this by using `right join`. You just need to set `how='right'`.***

![](right.png)
---

In [21]:
roll_no = pd.DataFrame({
    'roll_no' : [ 102, 103]
})
roll_no

roll_no
0      102
1      103

In [22]:
student_df

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

In [23]:
student_df.merge(roll_no, how='right', on='roll_no')

roll_no     name grade  marks      city
0      102  Aravind     B     15  Gurugram
1      103  Piyuesh     C      4  Gurugram

***Now, students participated in a placement drive and some of the students got placed and the details are given in the `student_selection` dataframe.***

---

In [24]:
student_selection = pd.DataFrame({
    'roll_no' : [102, 105, 101],
    'company' : ['ABC', 'XYZ', 'ABC'],
    'package (lpa)' : [ 8, 14.5, 11 ]
})
student_selection

roll_no company  package (lpa)
0      102     ABC            8.0
1      105     XYZ           14.5
2      101     ABC           11.0

***Now, we want to combine the `student_df` and `student_selection`. We can do this by using `outer/full join`. You need to set parameter `how = 'outer'`.***

![](outer.png)

---

In [25]:
student_df

roll_no     name grade  marks       city
0      102  Aravind     B     15   Gurugram
1      101    Rahul     B     15      Delhi
2      104  Prateek     A     20      Delhi
3      103  Piyuesh     C      4   Gurugram
4      105   Kartik     A     22  Hyderabad

In [26]:
student_df.merge(student_selection, how='outer')

roll_no     name grade  marks       city company  package (lpa)
0      101    Rahul     B     15      Delhi     ABC           11.0
1      102  Aravind     B     15   Gurugram     ABC            8.0
2      103  Piyuesh     C      4   Gurugram     NaN            NaN
3      104  Prateek     A     20      Delhi     NaN            NaN
4      105   Kartik     A     22  Hyderabad     XYZ           14.5

---

***Now, consider it was a pool placement drive and students from multiple colleges participated. The college `"ZU UNIVERSITY"` got the list of students selected for the job. You need to find out the details of the students who got selected from the college `"ZU UNIVERSITY"`***


---

In [27]:
student_df = pd.DataFrame({
    'college': ['ZU UNIVERSITY', 'ZU UNIVERSITY', 'ZU UNIVERSITY', 'ZU UNIVERSITY', 'ZU UNIVERSITY'],
    'roll_no': [ 102, 101, 104, 103, 105],
    'name' : ['Aravind', 'Rahul', 'Prateek', 'Piyuesh', 'Kartik'],
    'grade': ['B', 'B', 'A', 'C', 'A'],
    'marks': [ 15, 15, 20, 4, 22],
    'city' : ['Gurugram', 'Delhi', 'Delhi', 'Gurugram', 'Hyderabad']
})
student_df

college  roll_no     name grade  marks       city
0  ZU UNIVERSITY      102  Aravind     B     15   Gurugram
1  ZU UNIVERSITY      101    Rahul     B     15      Delhi
2  ZU UNIVERSITY      104  Prateek     A     20      Delhi
3  ZU UNIVERSITY      103  Piyuesh     C      4   Gurugram
4  ZU UNIVERSITY      105   Kartik     A     22  Hyderabad

In [28]:
pool = pd.DataFrame({
        'college': ['ZU UNIVERSITY', 'ZU UNIVERSITY', 'AB UNIVERSITY', 'ZU UNIVERSITY','AB UNIVERSITY'],
        'name' : ['Aravind', 'Rahul', 'Rahul', 'Prateek', 'Harsh'],
        'company' : ['ABC', 'XYZ', 'ABC', 'AEP', 'ABC'],
        'package (lpa)' : [ 8, 14.5, 11, 6, 6 ]
})
pool

college     name company  package (lpa)
0  ZU UNIVERSITY  Aravind     ABC            8.0
1  ZU UNIVERSITY    Rahul     XYZ           14.5
2  AB UNIVERSITY    Rahul     ABC           11.0
3  ZU UNIVERSITY  Prateek     AEP            6.0
4  AB UNIVERSITY    Harsh     ABC            6.0

---

***Now, we have 2 columns common `college` and `name` in both the dataframes. So, here we will use the `inner join`. You just need to set the parameters `how='inner'` and as we have 2 common columns therefore, set the parameter `on=['college','name']`.***


![](inner.png)

---

In [29]:
student_df.merge(pool, how='inner', on=['college', 'name'])

college  roll_no     name grade  marks      city company  \
0  ZU UNIVERSITY      102  Aravind     B     15  Gurugram     ABC   
1  ZU UNIVERSITY      101    Rahul     B     15     Delhi     XYZ   
2  ZU UNIVERSITY      104  Prateek     A     20     Delhi     AEP   

   package (lpa)  
0            8.0  
1           14.5  
2            6.0

---

### `NOW LET'S SOLVE THE PROBLEM`

---

We have another dataset `outlet_data.csv` in the dataset folder. It has column `Outlet_Identifier`, `Outlet_Establishment_Year`, `Outlet_Size` and `Outlet_Location_Type`.


Now, we have `Outlet_Identifier` in both the datasets and we need to combine them and get the rest of the variables in our dataset. So we will do a `left join` to merge the data frames. Let's see how?

---


#### `READ THE OUTLET DATA`

---

In [ ]:
### all merges with sample

In [30]:
previous_data = pd.read_csv('/content/drive/My Drive/Pandas_Concatenation/Concatenation data/outlet_size_concatenated_data.csv')

In [31]:
previous_data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  
0  144.1102            OUT046  Supermarket Type1          2187.1530  
1  119.6782            OUT046  Supermarket Type1          2145.2076  
2   56.3614            OUT046  Supermarket Type1          1547.3192  
3  230.5352            OUT035  Supermarket Type1          2748.4224  
4  196.5084            OUT035  Supermarket Type1          1587.2672

In [32]:
# read the outlet data
outlet_data = pd.read_csv('/content/drive/My Drive/Pandas_Concatenation/Concatenation data/outlet_data.csv')

In [33]:
# view the top rows of the data
outlet_data.head()

Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0            OUT013                       1987        High   
1            OUT018                       2009      Medium   
2            OUT019                       1985       Small   
3            OUT027                       1985      Medium   
4            OUT035                       2004       Small   

  Outlet_Location_Type  
0               Tier 3  
1               Tier 3  
2               Tier 1  
3               Tier 3  
4               Tier 2

---


***Use the merge function to set parameter `how = 'left'` for the left join and the set the on parameter as the common column name as `on='Outlet_Identifier'`.***


---

In [35]:
# merge the data
combined_data = previous_data.merge(outlet_data, how='left', on='Outlet_Identifier')

In [36]:
# view the data
combined_data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDA03        18.50          Regular         0.045464        Dairy   
1           FDS46        17.60          Regular         0.047257  Snack Foods   
2           FDP49         9.00          Regular         0.069089    Breakfast   
3           FDU02        13.35          Low Fat         0.102492        Dairy   
4           NCB30        14.60          Low Fat         0.025698    Household   

   Item_MRP Outlet_Identifier        Outlet_Type  Item_Outlet_Sales  \
0  144.1102            OUT046  Supermarket Type1          2187.1530   
1  119.6782            OUT046  Supermarket Type1          2145.2076   
2   56.3614            OUT046  Supermarket Type1          1547.3192   
3  230.5352            OUT035  Supermarket Type1          2748.4224   
4  196.5084            OUT035  Supermarket Type1          1587.2672   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  
0                       1997       Small               Tier 1  
1                       1997       Small               Tier 1  
2                       1997       Small               Tier 1  
3                       2004       Small               Tier 2  
4                       2004       Small               Tier 2

In [37]:
combined_data.shape

(6113, 12)

----

So, you can see that we got some extra variables in the dataset.

----


***Now, we have a sample list of Item_Identifiers in a separate file `item_identifier.csv`. We are required to provide all the details that we have related to that particular Item_Identifier. Let's see how can we do that with the help of a `RIGHT JOIN`***

---


#### `READ THE DATA`

---

In [38]:
# read the data
item_data = pd.read_csv('/content/drive/My Drive/Pandas_Concatenation/Concatenation data/item_idenifier.csv')

In [39]:
# view the top rows
item_data.head()

Item_Identifier
0           DRI51
1           FDL48
2           FDL38
3           FDF17
4           FDN56

In [40]:
item_data.shape

(100, 1)

***So, we have 100 Item_Identifiers and we need to provide the other details. We just need to use the merge function and set parameter `how='right'` and the common variable name is `Item_Identifier`.***

---

In [41]:
item_details = combined_data.merge(item_data, how='right',on='Item_Identifier')

In [42]:
item_details

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0             DRI51        17.25          Low Fat         0.042234   
1             DRI51        17.25          Low Fat         0.042414   
2             DRI51          NaN          Low Fat         0.042037   
3             FDL48        19.35          Regular         0.082251   
4             FDL48        19.35          Regular         0.082266   
..              ...          ...              ...              ...   
417           FDD57        18.10               LF         0.022381   
418           FDG41         8.84          Regular         0.076548   
419           FDG41         8.84          Regular         0.076681   
420           FDG41         8.84          Regular         0.076874   
421           NCM42         6.13          Low Fat         0.028366   

                 Item_Type  Item_MRP Outlet_Identifier        Outlet_Type  \
0                    Dairy  173.3764            OUT035  Supermarket Type1   
1                    Dairy  173.1764            OUT018  Supermarket Type2   
2                    Dairy  172.6764            OUT027  Supermarket Type3   
3             Baking Goods   48.7034            OUT035  Supermarket Type1   
4             Baking Goods   48.8034            OUT046  Supermarket Type1   
..                     ...       ...               ...                ...   
417  Fruits and Vegetables   93.6094            OUT013  Supermarket Type1   
418           Frozen Foods  109.5228            OUT035  Supermarket Type1   
419           Frozen Foods  110.7228            OUT049  Supermarket Type1   
420           Frozen Foods  109.9228            OUT018  Supermarket Type2   
421              Household  110.0912            OUT049  Supermarket Type1   

     Item_Outlet_Sales  Outlet_Establishment_Year Outlet_Size  \
0            2061.3168                       2004       Small   
1            4466.1864                       2009      Medium   
2            6183.9504                       1985      Medium   
3             534.6374                       2004       Small   
4             340.2238                       1997       Small   
..                 ...                        ...         ...   
417           476.0470                       1987        High   
418          1657.8420                       2004       Small   
419          1657.8420                       1999      Medium   
420          1547.3192                       2009      Medium   
421          1637.8680                       1999      Medium   

    Outlet_Location_Type  
0                 Tier 2  
1                 Tier 3  
2                 Tier 3  
3                 Tier 2  
4                 Tier 1  
..                   ...  
417               Tier 3  
418               Tier 2  
419               Tier 1  
420               Tier 3  
421               Tier 1  

[422 rows x 12 columns]